pkg/apis/core/v1/helper/helpers.go

```diff
@@ -315,34 +316,43 @@ func NodeSelectorRequirementKeysExistInNodeSelectorTerms(reqs []v1.NodeSelectorR
 // MatchNodeSelectorTerms checks whether the node labels and fields match node selector terms in ORed;
 // nil or empty term matches no objects.
 func MatchNodeSelectorTerms(
+	node *v1.Node,
 	nodeSelectorTerms []v1.NodeSelectorTerm,
-	nodeLabels labels.Set,
-	nodeFields fields.Set,
-) bool {
+) (bool, error) {
+	var errors []error
 	for _, req := range nodeSelectorTerms {
 		// nil or empty term selects no objects
 		if len(req.MatchExpressions) == 0 && len(req.MatchFields) == 0 {
 			continue
 		}
 
 		if len(req.MatchExpressions) != 0 {
-			labelSelector, err := NodeSelectorRequirementsAsSelector(req.MatchExpressions)
-			if err != nil || !labelSelector.Matches(nodeLabels) {
+			labelSelector, err := nodeSelectorRequirementsAsSelector(req.MatchExpressions)
+			if err != nil {
+				errors = append(errors, err)
+			}
+			if labelSelector == nil || !labelSelector.Matches(labels.Set(node.Labels)) {
 				continue
 			}
 		}
 
 		if len(req.MatchFields) != 0 {
-			fieldSelector, err := NodeSelectorRequirementsAsFieldSelector(req.MatchFields)
-			if err != nil || !fieldSelector.Matches(nodeFields) {
+			fieldSelector, err := nodeSelectorRequirementsAsFieldSelector(req.MatchFields)
+			if err != nil {
+				errors = append(errors, err)
+				continue
+			}
+			if fieldSelector == nil || !fieldSelector.Matches(fields.Set{
+				"metadata.name": node.Name,
```

Comment:
should we only include this is `node.Name` is present, or should we assume nodes passed in are valid, persisted nodes (the current PR would need to be changed to make that assumption correct)?

Reply:
I see your point now, yeah the access here could be a problem (like in the case of the storage code which I'm trying to update to a synthetic node object)

I think we should only make this check if `node.Name` is present. My reasoning for that is:
1. `node.Name` is the only valid field selector
2. any current users of this function will either a) have access to `node.Name` (by passing in the `nodeFields` param) and thus able to at least parse that into a Node literal as a temporary solution or b) not currently passing in anything and so don't care about this check